### ISSUE TITLE
Status of your repo

### ISSUE THREAD

**Issue by m-toman** [h-meru/Tacotron-WaveRNN#8]
Hi, just found your repo and saw you continued working on it.

What I did meanwhile was adapting the repo from https://github.com/geneing/WaveRNN-Pytorch
and tested implementing the Amazon version of WaveRNN (https://arxiv.org/abs/1811.06292).

I've noticed it's quite similar to yours
(here my messy forward: https://github.com/m-toman/tacorn/blob/master/wavernn/model.py#L160-L189)
except that they use:
- only one GRU layer and no residual connections
- no initial FC layer
- an additional conditioning network of a bidrectional GRU layer.
- 10 bit mulaw and they also seem to feed one-hot-encoding as input to the network instead of raw

I'm rather unhappy with the results till now, but I'm just at 100k steps, perhaps it will get better.

Are the (old) samples in your readme representative?

Nice work BTW ;)

**Comment by alexdemartos** [comment]
Hey @m-toman! Have you finally managed to get any good results with your Amazon version?

**Comment by m-toman** [comment]
Hi @alexdemartos 
Not really, at the moment I'm still getting the best results with this architecture here (2x 512er GRU with residual connection), 10 bit quantization or MoL (although MoL is much harder to train).

At the moment I'm trying again to use the Amazon-style preconditioning network implemented as:

`class MelConditionNetwork(nn.Module):
    def __init__(self, input_dim, rnn_dim, rnn_num_layers, rnn_dropout):
        super().__init__()
        self.input_dim = input_dim
        self.rnn_dim = rnn_dim
        self.rnn_num_layers = rnn_num_layers
        self.batch_norm = nn.BatchNorm1d(rnn_dim * 2)

        self.rnn = nn.GRU(input_size=input_dim, hidden_size=rnn_dim,
                          num_layers=rnn_num_layers, batch_first=True,
                          bidirectional=True, dropout=0.1)

    def forward(self, m):
        # TODO: do this once and modify upsample network
        # input is (batch_size, mel_dims, time), for the GRU we want (batch_size, time, mel_dims)
        m = m.transpose(1, 2)
        m, _ = self.rnn(m)
        # but batch norm again wants (N, C, L)
        m = m.transpose(1, 2)
        return self.batch_norm(m)
`

but again I get the impression that it only makes the training less stable and obviously slower. Testing on LJ with MoL, a bit unsure about gradient norm clipping (gradient norm is always around 3k).
Without the preconditioning network the loss drops below 5.5 after 1k epochs (of batch size 128)
With it it's still at 7 after 1k epochs.

Perhaps it does help when using data from lots of different speakers, tried it with more than 300k sentences from more than 100 different speakers but it never got to a good state yet. Also I'd prefer training from GTA mels (and for that I'd probably first have to train a large Tacotron multispeaker model...)

Generally it's annoying as hell :)

**Comment by alexdemartos** [comment]
Thanks for the detailed update! 

In my case I do have a large Tacotron-2 multi-speaker model up and running. I am now trying to get as good vocoding quality as possible. Right now I am trying with 10 bit mu-law model, but it seems it needs some time to converge...I am also trying to get MoL training to work (from erogol WaveRNN implementation)

**Comment by m-toman** [comment]
Personally every time I tried mulaw instead of "regular" 10-bit quantization I had a much harder time with the loss jumping around and exploding more often.

I'm still mostly experimenting with https://github.com/Rayhane-mamah/Tacotron-2 but I guess in the long run it might be better to switch to Mozilla TTS.
But until now I got more robust results with Rayhane's implementation, especially when adapting/fine-tuning a small dataset from a base model - it then usually gets good results after a couple steps. With Mozilla TTS I'm still struggling, but I guess this is mostly because of the dropout/batchnorm-thing.

Back to WaveRNN: I also found the noam scheme with the standard settings to not really fit, so I'm currently manually lower the learning rate whenever the loss starts to act funny.
But at some point it usually gets really hard to further get the loss down.

I've probably stopped this model 30 times already to adapt learning rate, gradient clipping and the sequence len ;) (that's MoL)
![image](https://user-images.githubusercontent.com/9433378/58165025-36016100-7c87-11e9-890b-a84646778500.png)

The 10-bit quant was much easier to train and did it to 2mio steps...

**Comment by alexdemartos** [comment]
Your loss peaks seem kind of regular (every 2k steps). I am wondering if this might be due to some outlier utterances in your corpora?

I am also using (my heavily modified) Rayhane-mamma's version (+GST, +speaker embeddings, +multi-lingual, etc), but I am also following close Mozilla TTS developments. If you try multi-speaker training, I strongly suggest you to try GST independently if you have or not speaker labels, and making the PreNet much narrower (64).

**Comment by m-toman** [comment]
Just a quick update, stopped the amazon preconditioning net training after a couple days because flattened out so much and I'm happier with a smaller network anyway.
Quickly compared MoL and 10-bit quantization in training again and yeah... 10 bit is much easier to train and required nearly no intervention from my side. Also none of those spikes (btw above it was every 2k epochs, not steps). Of course, MoL sounds nicer overall.

Currently just trying the vanilla Mozilla TTS training on LJ, with dropout prenet (and not batchnorm yet) and forward attention. I guess I should've disabled forward attention in the beginning, because at about 100k steps the alignment is still far from what Rayhanes Taco implementation gets after fewer steps.
Perhaps I'll also give that a shot later, don't want to use so many GPUs just for rather small variations...
